File: Algorithm/javascript/BAEKJOON/CLASS 2/2609.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdput,
});

let input = [];
rl.on("line", (line) => {
  if (!line) {
    rl.close();
  } else {
    input.push(line);
  }
}).on("close", () => {
  solution(input);
  process.exit();
});

function solution(input) {
  let numberArray = input[0].split(" ").map((i) => Number(i));
  let minNumber = Math.min(...numberArray);

  let sum = (numberArray) => {
    let sumNumbers = 0;
    for (let number of numberArray) {
      sumNumbers += number;
    }
    return sumNumbers;
  };

  let divisor = (number) => {
    let divisorArray = [1];
    for (let i = 2; i < parseInt(number / 2) + 1; i++) {
      if (number % i == 0) {
        divisorArray.push(i);
      }
    }
    divisorArray.push(number);
    return divisorArray;
  };

  let gcd = (numberArray) => {
    let commonDivisorArray = [];
    let divisorArray = divisor(minNumber);
    for (let i of divisorArray) {
      let count = 0;
      for (let number of numberArray) {
        if (number % i == 0) {
          count++;
        }
        if (count == numberArray.length) {
          commonDivisorArray.push(i);
        }
      }
    }
    let theGreatestCommonDivisor = Math.max(...commonDivisorArray);
    return theGreatestCommonDivisor;
  };

  let lcm = (numberArray) => {
    let theGreatestCommonDivisor = gcd(numberArray);
    let theLeastCommonMultiple = gcd(numberArray);
    for (let i = 0; i < numberArray.length; i++) {
      numberArray[i] = parseInt(numberArray[i] / theGreatestCommonDivisor);
    }
    while (sum(numberArray) > numberArray.length) {
      for (let j = Math.max(...numberArray); j >= 1; j--) {
        for (let k = 0; k < numberArray.length; k++) {
          if (numberArray[k] % j == 0) {
            numberArray[k] = parseInt(numberArray[k] / j);
            theLeastCommonMultiple = theLeastCommonMultiple * j;
          }
        }
      }
    }
    console.log(theLeastCommonMultiple);
  };
  console.log(gcd(numberArray));
  lcm(numberArray);
}

File: Algorithm/python/BAEKJOON/CLASS 2/2609.py
# TRY 1
'''
이 문제에서는 2개의 수의 최대공약수 최소공배수를 구하라고 했지만, 3개 이상에서 최대공약수 최소공배수 구하고 싶다.
'''
number_list = list(map(int, input().split(" ")))
min_number = min(number_list)

def divisor(number):
    divisor_list = [1]
    for i in range(2, number//2 + 1):
        if number % i == 0:
            divisor_list.append(i)
    divisor_list.append(number)
    return divisor_list

def gcd(number_list):
    common_divisor_list = []
    divisor_list = divisor(min_number)
    for i in  divisor_list:
        count = 0
        for number in number_list:
            if number % i == 0:
                count += 1
            if count == len(number_list):
                common_divisor_list.append(i)
    return common_divisor_list[-1]

def lcm(number_list):
    least_common_multiple = gcd(number_list)
    number_list = [int(number/gcd(number_list)) for number in number_list]
    while sum(number_list) > len(number_list):
        for i in range(max(number_list), 1, -1):
            for j in range(len(number_list)):
                if number_list[j] % i == 0:
                    number_list[j] //= i
                    least_common_multiple *= i
    return least_common_multiple

print(gcd(number_list))
print(lcm(number_list))